File: sdks/csharp/src/NovaBilling/Types/SubscriptionResponse.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using NovaBilling.Core;

namespace NovaBilling;

[Serializable]
public record SubscriptionResponse : IJsonOnDeserialized
{
    [JsonExtensionData]
    private readonly IDictionary<string, JsonElement> _extensionData =
        new Dictionary<string, JsonElement>();

    [JsonPropertyName("id")]
    public required string Id { get; set; }

    [JsonPropertyName("externalId")]
    public string? ExternalId { get; set; }

    [JsonPropertyName("customerId")]
    public required string CustomerId { get; set; }

    [JsonPropertyName("planId")]
    public required string PlanId { get; set; }

    [JsonPropertyName("previousPlanId")]
    public string? PreviousPlanId { get; set; }

    [JsonPropertyName("status")]
    public required SubscriptionResponseStatus Status { get; set; }

    [JsonPropertyName("currency")]
    public required string Currency { get; set; }

    [JsonPropertyName("billingTiming")]
    public required SubscriptionResponseBillingTiming BillingTiming { get; set; }

    [JsonPropertyName("currentPeriodStart")]
    public required string CurrentPeriodStart { get; set; }

    [JsonPropertyName("currentPeriodEnd")]
    public required string CurrentPeriodEnd { get; set; }

    [JsonPropertyName("cancelAt")]
    public string? CancelAt { get; set; }

    [JsonPropertyName("canceledAt")]
    public string? CanceledAt { get; set; }

    [JsonPropertyName("trialStart")]
    public string? TrialStart { get; set; }

    [JsonPropertyName("trialEnd")]
    public string? TrialEnd { get; set; }

    [JsonPropertyName("startedAt")]
    public required string StartedAt { get; set; }

    [JsonPropertyName("metadata")]
    public Dictionary<string, object?>? Metadata { get; set; }

    [JsonPropertyName("customer")]
    public SubscriptionCustomerResponse? Customer { get; set; }

    [JsonPropertyName("plan")]
    public SubscriptionPlanResponse? Plan { get; set; }

    [JsonPropertyName("createdAt")]
    public required string CreatedAt { get; set; }

    [JsonPropertyName("updatedAt")]
    public required string UpdatedAt { get; set; }

    [JsonIgnore]
    public ReadOnlyAdditionalProperties AdditionalProperties { get; private set; } = new();

    void IJsonOnDeserialized.OnDeserialized() =>
        AdditionalProperties.CopyFromExtensionData(_extensionData);

    /// <inheritdoc />
    public override string ToString()
    {
        return JsonUtils.Serialize(this);
    }
}

File: sdks/ruby/lib/novabilling/types/subscription_response.rb
# frozen_string_literal: true

module Novabilling
  module Types
    class SubscriptionResponse < Internal::Types::Model
      field :id, -> { String }, optional: false, nullable: false
      field :external_id, -> { String }, optional: true, nullable: false, api_name: "externalId"
      field :customer_id, -> { String }, optional: false, nullable: false, api_name: "customerId"
      field :plan_id, -> { String }, optional: false, nullable: false, api_name: "planId"
      field :previous_plan_id, -> { String }, optional: true, nullable: false, api_name: "previousPlanId"
      field :status, -> { Novabilling::Types::SubscriptionResponseStatus }, optional: false, nullable: false
      field :currency, -> { String }, optional: false, nullable: false
      field :billing_timing, -> { Novabilling::Types::SubscriptionResponseBillingTiming }, optional: false, nullable: false, api_name: "billingTiming"
      field :current_period_start, -> { String }, optional: false, nullable: false, api_name: "currentPeriodStart"
      field :current_period_end, -> { String }, optional: false, nullable: false, api_name: "currentPeriodEnd"
      field :cancel_at, -> { String }, optional: true, nullable: false, api_name: "cancelAt"
      field :canceled_at, -> { String }, optional: true, nullable: false, api_name: "canceledAt"
      field :trial_start, -> { String }, optional: true, nullable: false, api_name: "trialStart"
      field :trial_end, -> { String }, optional: true, nullable: false, api_name: "trialEnd"
      field :started_at, -> { String }, optional: false, nullable: false, api_name: "startedAt"
      field :metadata, -> { Internal::Types::Hash[String, Object] }, optional: true, nullable: false
      field :customer, -> { Novabilling::Types::SubscriptionCustomerResponse }, optional: true, nullable: false
      field :plan, -> { Novabilling::Types::SubscriptionPlanResponse }, optional: true, nullable: false
      field :created_at, -> { String }, optional: false, nullable: false, api_name: "createdAt"
      field :updated_at, -> { String }, optional: false, nullable: false, api_name: "updatedAt"
    end
  end
end
